File: Problems/Algorithms/947. Most Stones Removed with Same Row or Column/MostStonesRemoved.java
class Solution {
    public int removeStones(int[][] stones) {
        int size = stones.length;
        UnionFind uf = new UnionFind(size);
        for (int i = 0; i < size; i++) {
            for (int j = i+1; j < size; j++) {
                if (stones[i][0] == stones[j][0] || stones[i][1] == stones[j][1]) {
                    uf.union(i, j);
                }
            }
        }
        
        int roots = 0;
        HashSet<Integer> used = new HashSet<>();
        for (int i = 0; i < size; i++) {
            int root = uf.find(i);
            if (!used.contains(root)) {
                used.add(root);
                roots++;
            }
        }
        
        return size - roots;
    }
    
    private class UnionFind {
        private int[] root;
        private int[] rank;
        
        private UnionFind(int size) {
            root = new int[size];
            rank = new int[size];
            for (int i = 0; i < size; i++) {
                root[i] = i;
                rank[i] = 1;
            }
        }
        
        private int find(int x) {
            if (x == root[x]) {
                return x;
            }
            return root[x] = find(root[x]);
        }
        
        private void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            
            if (rootX != rootY) {
                if (rank[rootX] > rank[rootY]) {
                    root[rootY] = rootX;
                } else if (rank[rootX] < rank[rootY]) {
                    root[rootX] = rootY;
                } else {
                    root[rootY] = rootX;
                    rank[rootX] += 1;
                }
            }
        }
    }
}

File: Problems/Algorithms/947. Most Stones Removed with Same Row or Column/most_stones_removed.py
from typing import List


class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.set_count = n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        fx, fy = self.find(x), self.find(y)
        if fx == fy:
            return False

        if self.size[fx] < self.size[fy]:
            fx, fy = fy, fx
        self.parent[fy] = fx
        self.size[fx] += self.size[fy]
        self.set_count -= 1
        return True

class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        n = len(stones)
        uf = UnionFind(n)
        for i in range(n):
            for j in range(i + 1, n):
                if stones[i][0] == stones[j][0] or stones[i][1] == stones[j][1]:
                    uf.union(i, j)

        used = {}
        for i in range(n):
            root = uf.find(i)
            if root not in used:
                used[root] = 1
        
        return n - len(used)
